src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs

```diff
@@ -855,15 +855,23 @@ public void Close()
                 {
                     if (_standardOutput != null && (_outputStreamReadMode == StreamReadMode.AsyncMode || _outputStreamReadMode == StreamReadMode.Undefined))
                     {
-                    _standardOutput.Close();
+                        if (_outputStreamReadMode == StreamReadMode.AsyncMode)
+                        {
+                            _output.CancelOperation();
+                        }
+                        _standardOutput.Close();
```

Comment:
We will not target race condition between `Close` and parallel `AsyncStreamReader` operations with current tests.
Looking at the [`AsyncStreamReader` code](https://github.com/dotnet/coreclr/blob/2142989645f124c8fbaa06acd06d2bb960168b1a/src/mscorlib/src/System/IO/StreamReader.cs#L878-L879), we may at minimum get `ObjectDisposedException` after calling `Close` on the underlying `_standardOutput`/`_standardError` stream. `AsyncStreamReader` does not seem to swallow that one [here](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs#L101-L115).

It would be good IMO to manually stress test the race conditions - e.g. modify the added test to write all it can (for a period of time) and call `Close` in the middle of it. And run the test in a loop for a day.

Reply:
My understanding is calling `_output.CancelOperation()` should entirely cancel the async reading, and prevent the callback from getting called until `_output.BeginReadLine()` is called again, preventing the race condition on that? If I'm wrong about that definitely let me know.